library/psa_crypto.c

```diff
@@ -2853,6 +2855,140 @@ psa_status_t psa_mac_verify_finish( psa_mac_operation_t *operation,
 /* Asymmetric cryptography */
 /****************************************************************/
 
+psa_status_t psa_sign_message( psa_key_id_t key,
+                               psa_algorithm_t alg,
+                               const uint8_t * input,
+                               size_t input_length,
+                               uint8_t * signature,
+                               size_t signature_size,
+                               size_t * signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot;
+    size_t hash_length;
+    uint8_t hash[PSA_HASH_MAX_SIZE];
+
+    *signature_length = 0;
+
+    if( ! PSA_ALG_IS_SIGN_MESSAGE( alg ) )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
```

Comment:
The code beyond this point is only valid when `PSA_ALG_IS_SIGN_HASH( alg )`. Otherwise the hash-then-sign method is not valid. Currently, only one algorithm is declared which is not a sign-hash algorithm: `PSA_ALG_PURE_EDDSA`. And it is not yet implemented in Mbed TLS, so it can't be mentioned here, but we must not attempt to do the hash-then-sign method for it.

Also goes for verify.